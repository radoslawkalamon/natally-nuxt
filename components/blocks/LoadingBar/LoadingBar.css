.loading-bar {
  bottom: 0;
  box-sizing: border-box;
  left: 0;
  position: fixed;
  width: 100%;
  z-index: 9;
}

.loading-bar .loading-bar__inner {
  border-radius: var(--border-radius) var(--border-radius) 0 0;
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity var(--time-animation);
}

.fade-enter,
.fade-leave-to {
  opacity: 0;
}
