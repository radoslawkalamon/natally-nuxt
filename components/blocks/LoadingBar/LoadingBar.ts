import Vue from 'vue'

type BlocksLoadingBarData = {
  progress: number;
  shallShow: boolean;
  timeout: number | NodeJS.Timeout;
}

export default Vue.extend({
  name: 'BlocksLoadingBar',
  components: {
    ComponentsProgressBar: () => import(/* webpackChunkName: "components-progress-bar" */'@/components/components/ProgressBar/ProgressBar.vue')
  },
  data (): BlocksLoadingBarData {
    return {
      shallShow: false,
      progress: 1,
      timeout: 0
    }
  },
  methods: {
    finish (): void {
      this.setProgress(100)
      this.setTimeout(this.hideLoadingBar)
    },
    increase (n: number): void {
      this.setProgress(n)
    },
    start (): void {
      this.clearTimeout()
      this.showLoadingBar()
      this.setProgress(1)
    },
    setProgress (n: number): void {
      this.progress = n
    },
    hideLoadingBar (): void {
      this.shallShow = false
    },
    showLoadingBar (): void {
      this.shallShow = true
    },
    clearTimeout (): void {
      clearTimeout(Number(this.timeout))
    },
    setTimeout (callback: Function): void {
      this.timeout = setTimeout(() => callback(), 250)
    }
  }
})
