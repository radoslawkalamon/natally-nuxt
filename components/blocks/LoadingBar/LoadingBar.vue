<template>
  <transition name="fade">
    <div
      v-if="shallShow"
      aria-hidden="true"
      class="loading-bar"
    >
      <ComponentsProgressBar
        class="loading-bar__inner"
        :width="progress"
      />
    </div>
  </transition>
</template>

<script src="./LoadingBar.ts" lang="ts"></script>
<style src="./LoadingBar.css" scoped></style>
