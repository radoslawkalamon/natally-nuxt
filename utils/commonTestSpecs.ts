import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
