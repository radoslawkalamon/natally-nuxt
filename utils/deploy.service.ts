/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
