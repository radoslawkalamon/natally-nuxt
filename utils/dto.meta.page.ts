export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
