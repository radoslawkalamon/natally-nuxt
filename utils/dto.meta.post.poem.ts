import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
