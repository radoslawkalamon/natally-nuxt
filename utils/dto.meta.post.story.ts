import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostStoryConstructor = DTOMetaPageConstructor & {
  audiobookId?: string;
  imageCover: string;
  imageCover2x: string;
  imageOpenGraph: string
  timeReading: number;
}

export class DTOMetaPostStory extends DTOMetaPage {
  readonly audiobookId: string;
  readonly imageCover: string;
  readonly imageCover2x: string;
  readonly imageOpenGraph: string;
  readonly timeReading: number;

  constructor ({
    audiobookId = '0',
    createdAt,
    description,
    imageCover,
    imageCover2x,
    imageOpenGraph,
    path,
    timeReading,
    title,
    updatedAt
  }: DTOMetaPostStoryConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.audiobookId = audiobookId
    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageCover2x = `${this.imagePath}${imageCover2x}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
    this.timeReading = timeReading
  }

  get imagePath (): string {
    return '/images/'
  }
}
