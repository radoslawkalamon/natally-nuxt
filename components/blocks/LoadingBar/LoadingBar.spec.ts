import LoadingBar from '@/components/blocks/LoadingBar/LoadingBar.vue'
import { shallPassIntegrationSanityTest, shallPassUnitSanityTest } from '@/devtools/jest.common.spec'
import { createDefaultOptionsFactory, createUnitTestWrapper } from '@/devtools/jest.common.spec.utils'

jest.useFakeTimers()

const defaultOptionsFactory = createDefaultOptionsFactory({
  stubs: ['ComponentsProgressBar']
})

describe('Blocks / Loading Bar', () => {
  describe('Unit', () => {
    shallPassUnitSanityTest({
      component: LoadingBar,
      options: defaultOptionsFactory()
    })

    test('shall show progress bar', async () => {
      const wrapper = await createUnitTestWrapper({
        component: LoadingBar,
        options: defaultOptionsFactory()
      })
      // @ts-ignore: No component typing :(
      wrapper.vm.start()
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      wrapper.destroy()
    })

    test('shall increase progress bar width', async () => {
      const wrapper = await createUnitTestWrapper({
        component: LoadingBar,
        options: defaultOptionsFactory({
          data () {
            return {
              shallShow: true
            }
          }
        })
      })
      // @ts-ignore: No component typing :(
      wrapper.vm.increase(40)
      await wrapper.vm.$nextTick()

      expect(wrapper.html()).toMatchSnapshot()

      wrapper.destroy()
    })

    test('shall progress bar width be 100 & hide after timeout', async () => {
      const wrapper = await createUnitTestWrapper({
        component: LoadingBar,
        options: defaultOptionsFactory({
          data () {
            return {
              shallShow: true,
              progress: 40
            }
          }
        })
      })
      // @ts-ignore: No component typing :(
      wrapper.vm.finish()
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
      jest.advanceTimersByTime(2000)
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()

      wrapper.destroy()
    })

    test('shall loading bar be visible after finish() & start()', async () => {
      const wrapper = await createUnitTestWrapper({
        component: LoadingBar,
        options: defaultOptionsFactory({
          data () {
            return {
              shallShow: true,
              progress: 40
            }
          }
        })
      })
      // @ts-ignore: No component typing :(
      wrapper.vm.finish()
      jest.advanceTimersByTime(100)
      await wrapper.vm.$nextTick()
      // @ts-ignore: No component typing :(
      wrapper.vm.start()
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()

      wrapper.destroy()
    })
  })

  describe('Integration', () => {
    shallPassIntegrationSanityTest({
      component: LoadingBar,
      options: {
        data () {
          return {
            shallShow: true
          }
        }
      }
    })
  })
})
