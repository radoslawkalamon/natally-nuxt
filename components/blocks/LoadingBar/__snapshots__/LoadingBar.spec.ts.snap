// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocks / Loading Bar Integration shall pass integration sanity test 1`] = `
"<transition-stub name="fade">
  <div aria-hidden="true" class="loading-bar">
    <div class="progress-bar loading-bar__inner" style="width: 1.00%;"></div>
  </div>
</transition-stub>"
`;

exports[`Blocks / Loading Bar Unit shall increase progress bar width 1`] = `
"<transition-stub name="fade">
  <div aria-hidden="true" class="loading-bar">
    <componentsprogressbar-stub width="40" class="loading-bar__inner"></componentsprogressbar-stub>
  </div>
</transition-stub>"
`;

exports[`Blocks / Loading Bar Unit shall loading bar be visible after finish() & start() 1`] = `
"<transition-stub name="fade">
  <div aria-hidden="true" class="loading-bar">
    <componentsprogressbar-stub width="1" class="loading-bar__inner"></componentsprogressbar-stub>
  </div>
</transition-stub>"
`;

exports[`Blocks / Loading Bar Unit shall pass unit sanity test 1`] = `
"<transition-stub name="fade">
  <!---->
</transition-stub>"
`;

exports[`Blocks / Loading Bar Unit shall progress bar width be 100 & hide after timeout 1`] = `
"<transition-stub name="fade">
  <div aria-hidden="true" class="loading-bar">
    <componentsprogressbar-stub width="100" class="loading-bar__inner"></componentsprogressbar-stub>
  </div>
</transition-stub>"
`;

exports[`Blocks / Loading Bar Unit shall progress bar width be 100 & hide after timeout 2`] = `
"<transition-stub name="fade">
  <!---->
</transition-stub>"
`;

exports[`Blocks / Loading Bar Unit shall show progress bar 1`] = `
"<transition-stub name="fade">
  <div aria-hidden="true" class="loading-bar">
    <componentsprogressbar-stub width="1" class="loading-bar__inner"></componentsprogressbar-stub>
  </div>
</transition-stub>"
`;
